examples/erc20_gas/src/main.rs

```diff
@@ -0,0 +1,327 @@
+            .storage
+            .get(&treasury_balance_slot)
+            .expect("Treasury balance slot not found")
+            .present_value();
+
+        token_account.data.storage.insert(
+            treasury_balance_slot,
+            EvmStorageSlot::new_changed(
+                treasury_balance,
+                treasury_balance.saturating_sub(refund_amount),
+            ),
+        );
+
+        let caller_balance_slot: U256 = keccak256((caller, U256::from(3)).abi_encode()).into();
+        let caller_balance = token_account
+            .storage
+            .get(&caller_balance_slot)
+            .expect("Caller balance slot not found")
+            .present_value();
+
+        token_account.data.storage.insert(
+            caller_balance_slot,
+            EvmStorageSlot::new_changed(
+                caller_balance,
+                caller_balance.saturating_add(refund_amount),
+            ),
+        );
+
+        Ok(())
+    });
+
+    handler.post_execution.reward_beneficiary = Arc::new(|ctx, gas| {
+        let beneficiary = *ctx.evm.env.block.coinbase();
+        let gas_price = ctx.evm.env.effective_gas_price();
+        let reward = gas_price * U256::from(gas.spent() - gas.refunded() as u64);
+
+        let token_account = ctx
+            .evm
+            .inner
+            .journaled_state
+            .load_account(TOKEN, &mut ctx.evm.inner.db)
+            .map_err(EVMError::Database)?;
+
+        // Transfer reward from treasury to beneficiary
+        let treasury_balance_slot: U256 = keccak256((TREASURY, U256::from(3)).abi_encode()).into();
+        let treasury_balance = token_account
+            .storage
+            .get(&treasury_balance_slot)
+            .expect("Treasury balance slot not found")
+            .present_value();
+
+        token_account.data.storage.insert(
+            treasury_balance_slot,
+            EvmStorageSlot::new_changed(treasury_balance, treasury_balance.saturating_sub(reward)),
+        );
+
+        let beneficiary_balance_slot: U256 =
+            keccak256((beneficiary, U256::from(3)).abi_encode()).into();
+        let beneficiary_balance = token_account
+            .storage
+            .get(&beneficiary_balance_slot)
+            .expect("Beneficiary balance slot not found")
+            .present_value();
+
+        token_account.data.storage.insert(
+            beneficiary_balance_slot,
+            EvmStorageSlot::new_changed(
+                beneficiary_balance,
+                beneficiary_balance.saturating_add(reward),
+            ),
+        );
+
+        Ok(())
+    });
+
+    handler.validation.tx_against_state = Arc::new(|ctx| {
+        let token_account = ctx
+            .evm
+            .inner
+            .journaled_state
+            .load_account(TOKEN, &mut ctx.evm.inner.db)
+            .map_err(EVMError::Database)?;
+
+        validate_tx_against_account::<EvmWiringT, SPEC>(
+            token_account.data,
```

Comment:
This is not exactly correct, validation needs to be done for erc20 balance (that is storage value), not on token_account balance (that is account eth balance).